folding-schemes/src/folding/mova/nifs.rs

```diff
@@ -0,0 +1,457 @@
+use super::{CommittedInstance, InstanceWitness, Witness};
+use crate::arith::r1cs::R1CS;
+use crate::commitment::CommitmentScheme;
+use crate::folding::mova::pointvsline::{
+    PointVsLine, PointVsLineProof, PointvsLineEvaluationClaim,
+};
+use crate::folding::nova::nifs::NIFS as Nova;
+use crate::transcript::Transcript;
+use crate::utils::mle::dense_vec_to_dense_mle;
+use crate::utils::vec::{vec_add, vec_scalar_mul};
+use crate::Error;
+use ark_crypto_primitives::sponge::Absorb;
+use ark_ec::{CurveGroup, Group};
+use ark_ff::PrimeField;
+use ark_poly::MultilinearExtension;
+use ark_std::log2;
+use std::marker::PhantomData;
+
+/// Implements the Non-Interactive Folding Scheme described in section 4 of
+/// [Mova](https://eprint.iacr.org/2024/1220.pdf)
+/// `H` specifies whether the NIFS will use a blinding factor
+pub struct NIFS<
+    C: CurveGroup,
+    CS: CommitmentScheme<C, H>,
+    T: Transcript<C::ScalarField>,
+    const H: bool = false,
+> {
+    _c: PhantomData<C>,
+    _cp: PhantomData<CS>,
+    _ct: PhantomData<T>,
+}
+
+pub struct Proof<C: CurveGroup> {
+    pub h_proof: PointVsLineProof<C>,
+    pub mleE1_prime: C::ScalarField,
+    pub mleE2_prime: C::ScalarField,
+    pub mleT: C::ScalarField,
+}
+
+impl<C: CurveGroup, CS: CommitmentScheme<C, H>, T: Transcript<C::ScalarField>, const H: bool>
+    NIFS<C, CS, T, H>
+where
+    <C as Group>::ScalarField: Absorb,
+{
+    // Just a wrapper for Nova compute_T (compute cross-terms T) since the process is the same
+    pub fn compute_T(
+        r1cs: &R1CS<C::ScalarField>,
+        u1: C::ScalarField,
+        u2: C::ScalarField,
+        z1: &[C::ScalarField],
+        z2: &[C::ScalarField],
+    ) -> Result<Vec<C::ScalarField>, Error> {
+        Nova::<C, CS, H>::compute_T(r1cs, u1, u2, z1, z2)
+    }
+
+    // Protocol 7 - point 3 (16)
+    pub fn fold_witness(
+        a: C::ScalarField,
+        w1: &Witness<C>,
+        w2: &Witness<C>,
+        T: &[C::ScalarField],
+    ) -> Result<Witness<C>, Error> {
+        let a2 = a * a;
+        let E: Vec<C::ScalarField> = vec_add(
+            &vec_add(&w1.E, &vec_scalar_mul(T, &a))?,
+            &vec_scalar_mul(&w2.E, &a2),
+        )?;
+        let W: Vec<C::ScalarField> =
+            w1.W.iter()
+                .zip(&w2.W)
+                .map(|(i1, i2)| *i1 + (a * i2))
+                .collect();
+
+        let rW = w1.rW + a * w2.rW;
+        Ok(Witness::<C> { E, W, rW })
+    }
+
+    // Protocol 7 - point 3 (15)
+    pub fn fold_committed_instance(
+        a: C::ScalarField,
+        ci1: &CommittedInstance<C>,
+        ci2: &CommittedInstance<C>,
+        rE_prime: &[C::ScalarField],
+        mleE1_prime: &C::ScalarField,
+        mleE2_prime: &C::ScalarField,
+        mleT: &C::ScalarField,
+    ) -> Result<CommittedInstance<C>, Error> {
+        let a2 = a * a;
```

Comment:
This change makes sense to me but I think it would make more sense if we switched all the usage of 2 after the letter to symbolise square (e.g. Mova's `fold_witness`, Nova's `fold_witness` and `fold_committed_instance` etc.)

Reply:
in this case I think there is no space for confusion, since it's clear that `a2=a*a`, and the usage is only right at the following line.